### ISSUE TITLE
Vue.js devtools and Cookiebot by Usercentrics

### ISSUE THREAD

**Issue by thommyhh** [vuejs/devtools#740]
This is kind of strange behaviour and a bit difficult to describe, but I try my best.

I've installed the `Vue.js devtools` extension in Google Chrome (https://chromewebstore.google.com/detail/vuejs-devtools/nhdogjmejiglipccpnnnanhbledajbpd). When visiting websites, that use the Cookiebot by Usercentrics, and do not consent to all cookies, some (maybe all) javascript of the website stops working. Sometimes I can't even click on the main menu items.

When either consenting to all types of cookies or disabling the Vue.js devtools the site works as expected. I observed this behaviour on at least the following two sites:
* https://evaerk.dk/
* https://jannieoestergaard.dk/

**Comment by alexzhang1030** [comment]
Hey, if you don't open chrome devtools, we are not injecting any JS code into your extension host.

I'm not sure it's related to vue-devtools., I will try to inspect it when I have some free time.

**Comment by hognevevle** [comment]
We experience the same issue on https://www.webhuset.no/ . Whenever the Vue.js devtools extension is enabled, the following errors occur:

- Initial transitions do not fire on the site (basicly all the main content gets transitioned in, meaning it never appears unless you start scrolling)
- More troublesome: hover states do not fire as intended on the "Hosting" and "Logg inn" main menu items, causing their related dropdowns to not work. 

I also started seeing this the same time as we introduced the UserCentrics Cookiebot. It happens even without the dev tools being open at all, all it takes is for the extension to be enabled in Chrome.

**Comment by pdfrod** [comment]
Just ran into this issue today when trying to open https://koto.studio/work/amazon. The page simply doesn't work for me if I have Vue DevTools enabled (even if I accept all cookies and have the Chrome DevTools closed).

I did some debugging and this is what I observed:

- the [Cookiebot script](https://consent.cookiebot.com/uc.js) installs a MutationObserver that tracks all `script` tags injected in the page. I haven't delved much into why it's doing that, but I suspect that it's trying to delay all other scripts initialization until the user responds to the cookie banner;
- when Vue DevTools is enabled, the script `<script src="chrome-extension://nhdogjmejiglipccpnnnanhbledajbpd/build/detector-exec.js" type="text/plain"></script>` is injected in the page (I presume this is a script injected by Vue DevTools extension to detect if the page uses Vue);
- the Cookiebot code detects the `<script>` tag mentioned above, but seems to get a bit confused by it, because although the script runs in the page, it doesn't belong to the page (e.g., when trying to get the `parentNode` of the script tag, it gets `null`). This confusion is what I think makes Cookiebot go off the rails.

Other than not injecting any scripts on the page, I'm not sure there's anything that could be done in Vue Devtools to fix the issue. I think it's on Cookiebot to better handle scripts that come from browser extensions.

**Comment by andykillen** [comment]
@pdfrod Have found the same problem with Cookie Bot today.  It makes our site look like a white screen... I've just tried raising a ticket with usercentrics to see if there is any way around this, i.e. exclude scripts src's that start chrome-extension:// from its clutches.   Will report back here I have any joy.

**Comment by andykillen** [comment]
@pdfrod No response from usercentrics on the ticket, but was chatting with a GTM/cookiebot implementor and he suggested not using GTM to load cookie bot via their template, but to move to inserting the cookie bot scripts in the HEAD before the GTM script, and also using the manual blocking version of their script rather than the autoblocker.js that is used as standard. 

It means that we will have to manage our own script loading based on the cookie banner popup and its events, but will remove the problem of any extension that inserts into the page being broken.

**Comment by pdfrod** [comment]
Thanks for bringing the issue to Usercentrics @andykillen!

I don't use Cookiebot myself, but hopefully your tips will be useful to people that do use it.

**Comment by FeBe95** [comment]
I did some debugging and tried to reconstruct why this is happening. I learned a lot about the page lifecycle. I am using the Vue Devtools (v7.7.7) browser extension for Firefox, but the same should apply to the Chrome extension.

And yes, this took a while to put together, but hopefully it helps a fellow dev or two understand what’s actually going on here. Even while writing this up, I was still uncovering new details in how this actually works. 😅 


> [!NOTE]
> TL;DR
> 1. The Cookiebot script blocks script execution on page load using various DOM tricks.
> 2. It later tries to re-insert the blocked scripts into the DOM, so they can actually execute this time.
> 3. When `detector.js` is injected, it removes itself immediately after loading.
> 4. This is desired, since normally the script has already executed by then.
> 5. However, the Cookiebot script fails to re-insert the cached version because it cannot find the original script in the DOM anymore, making it impossible to determine its parent for `parent.insertBefore(new, original)`.
> 6. This error causes the re-insertion process to end prematurely, meaning that subsequent scripts are not inserted. At this stage, the page might not be rendered properly.

> [!IMPORTANT]
> ## Solution for Vue Devtools
> 
> There are two independent solutions to this issue that can be implemented on Vue Devtools' side, until the Cookiebot team has fixed it on their side. I’ve tested both solutions using Chrome’s “Load Unpacked Extension” tool, and they work equally well.
> 
> ### 👉 Solution 1
> Delete the `detector.onload` listener that removes the script tag after loading. However, I'm not sure if removing the script after execution was implemented as a precautionary measure, or if deleting the listener could cause unintended side effects.
> ```js
> // detector.onload = () => {
> //   detector.parentNode!.removeChild(detector)
> // }
> ```
> 
> ### 👉 Solution 2
> Add the Cookiebot-specific `data-cookieconsent` attribute before appending the script to the DOM:
> ```js
> detector.setAttribute('data-cookieconsent', 'ignore');
> // or:
> detector.dataset.cookieconsent = 'ignore';
> ```

> [!TIP]
> ## Solution for Cookiebot
> Again, two solutions exist. The first one is a simple yet effective way, so that the Vue Devtools and other extensions will work again. However, strictly speaking, any script could potentially remove itself from the DOM after loading. That's why the second solution exists.
>
> ### Solution 1
> Skip `<script>` tags if their `src` attribute starts with `moz-extension://` or `chrome-extension://`.
>
> ### Solution 2
> Improve the dequeuing mechanism, so that it simply skips the current script if no `parentNode` is found.


## Long Version

For those who are interested, here are my detailed findings. I cannot guarantee that all of this is 100% correct, but I tried my best to reconstruct what is actually happening. Have fun reading. 😊

When you load one of the affected pages, the following happens, in chronological order:
1. First, the browser loads and parses the DOM.

2. All scripts in the `<head>` element are loaded and executed in order.

3. The Cookiebot script should be placed as the first script in the `<head>` element, as stated in the [documentation](https://support.cookiebot.com/hc/en-us/articles/360009074960-Automatic-cookie-blocking#:~:text=Make%20sure%20that%20the%20Cookiebot%20script%20is%20the%20very%20first%20script%20to%20load%20on%20your%20website.):
    > Make sure that the Cookiebot script is the very first script to load on your website.

4. Therefore, the Cookiebot script [`uc.js`](https://consent.cookiebot.com/uc.js) is loaded and executed first. I put a prettified version of this script in a gist for you:
    → https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a768.

5. When the script is initialized ([gist#L2023](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L2023)) it checks for the presence of the cookie-consent cookie ([gist#L2359-L2360](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L2359-L2360)).
 
6. If consent has not been given (yet) and autoblocking is enabled (via the `data-blockingmode="auto"` attribute or `&blockingmode=auto` URL parameter), the script starts a somewhat complex mechanism to prevent scripts from executing automatically.
    **Note:** Most ad blockers (e.g. uBlock Origin) will hide the consent banner. Therefore, consent can never be given.

7. Back to the Cookiebot script. The previously mentioned mechanism starts off by creating a `new MutationObserver()` object that listens to subtree changes of the `<html>` document element ([gist#L2497](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L2497), [gist#L988](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L988)).
    **Note:** At this point the DOM is not yet fully loaded, because the synchronous `uc.js` script tag is blocking further DOM parsing.

8. The `mutationHandler()` function ([gist#L1012](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1012)) tracks all mutations and "postpones" node insertions ([gist#L1440](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1440)). Mutations that insert new `<script>` elements are handled separately ([gist#L1056](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1056)).
    **Note:** Postponing script insertions ensures that the scripts can be evaluated if they should be autoblocked from executing (e.g. marketing scripts) before being executed.

9. The actual "postponing" logic of a `<script>` element consists of:
    1. Changing the script type from `type="text/javascript"` (or `type="application/javascript"`) to `type="text/plain"` to prevent script execution in the browser ([gist#1456](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1456))
    2. Adding an event listener to the `beforescriptexecute` DOM event that calls `event.preventDefault()`, just to make sure.
        **Note:** This is only supported in Firefox ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/Element/beforescriptexecute_event#browser_compatibility)).
    3. In order of appearance, adding the script nodes to either the `nonAsyncMutations` or `deferMutations` array, depending on the presence of the element's [`defer`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script#defer) or [`async`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script#async) attribute ([gist#L1467-L1471](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1467-L1471)).

10. Now, the browser parses the next DOM tags after the Cookiebot script tag. Just like a "man-in-the-middle", the mutation observer transforms all scripts to `text/plain`. The browser inserts the transformed scripts and loads the source files from the servers.

11. Once this process is done, we are at the `document_idle` point in time. The browser has successfully parsed and built the DOM. To be more specific: All elements in `<head>` and `<body>` exist, all `<script>` element source files are loaded, but not executed due to `text/plain`. Also, the mutation observer is still observing.

12. Now (at `document_idle`), Vue DevTools injects its [`detector.js`](https://github.com/vuejs/devtools/blob/main/packages/firefox-extension/src/detector.ts) script. It is injected by the [`devtools-overlay.js`](https://github.com/vuejs/devtools/blob/main/packages/firefox-extension/src/devtools-overlay.ts) helper, using the WebExtension manifest ([manifest.json#L29-L35](https://github.com/vuejs/devtools/blob/d36712f7a382a2bf61609dc2c797c65338822f51/packages/firefox-extension/manifest.json#L29-L35)). The browser adds the script to the DOM, the mutation observer does the usual "script postponing" by setting it to `type="text/plain"` and adding it to the `nonAsyncMutations` array.

13. Afterwards, the mutation handler is done processing. The two arrays will now look something like this:
    - `nonAsyncMutations`:
        - All `<script>` elements of the `<head>` element
        - All `<script>` elements of the `<body>` element
        - The `<script src="moz-extension://<UUID>/dist/detector.js" type="text/plain">` element
    - `deferMutations`:
        - All `<script>` elements that used the `defer` attribute
        - All `<script>` elements that used the `async` attribute

14. Next, the `onload` event of the `detector.js` script element is fired, which triggers a function that immediately removes the script element from the DOM tree again ([devtools-overlay.ts#L8-L10](https://github.com/vuejs/devtools/blob/d36712f7a382a2bf61609dc2c797c65338822f51/packages/chrome-extension/src/devtools-overlay.ts#L8-L10)).

15. Now, the browser fires `load` event of the `window` object. This will stop the mutation observer ([gist#L2081](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L2081) and start processing the postponed mutations [gist#L2603](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L2603)).

16. First, all non-script mutations are processed ([gist#L1105](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1005)), followed by the separately stored `<script>` elements ([gist#L1006](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1006)).

17. The stored scripts are processed one-by-one, using a FIFO-like queuing system, implemented in the `dequeueNonAsyncScripts()` function ([gist#L1193](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1193)). The array `nonAsyncMutations` is processed first, followed by the `deferMutations` array ([gist#L1251](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1251)).

18. The first `<script type="text/plain">` element is loaded into the queue. To ensure that the scripts are executed in the correct order, a recursive call to `dequeueNonAsyncScripts()` is attached as an `onload` event listener ([gist#L1230](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1230)). This means that the queue will wait for the current script to be inserted into the DOM and finished executing, before it continues processing the next script tag. If all scripts in `nonAsyncMutations` are processed, the queue loads all items from `deferMutations` and prcesses them the same way ([gist#L1251](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1251)).

19. For a script tag to execute again, it needs to be inserted into the DOM. Simply changing the `type` attribute isn't enough. That's why the Cookiebot script clones the original script tag that it has previously stored in one of the arrays ([gist#L1220](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1220)) and changes its `type` attribute back to `text/javascript` ([gist#L1129](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1129)).

20. Now comes the **critical** part: The cloned node insertion.
    1. The Cookiebot script tries to insert the cloned script tag into the DOM ([gist#L1239-L1246](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1239-L1246)).
    2. In order to do this, the browser requires a parent element into which the script can be inserted as a child ([gist#L1219](https://gist.github.com/FeBe95/85e0b1720c66c45830ff977880887a76#file-uc-js-L1219)).
    3. For most scripts, this works perfectly fine. The script is inserted, the `onload` listener triggers the next queue item to be processed.

21. However, the story is different for the `detector.js` script. Between building the `nonAsyncMutations` array and the queue processing, the original script element was removed by the `detector.onload` callback.

22. This means that `script.parentNode` of the cached script element is now `null` and the Cookiebot script fails to insert the node back into the DOM.

23. Consequently, the `script.onload` event never fires, and the queue is blocked indefinitely. There is no retry or skip mechanism of any sort.

24. **Why does this cause some pages to be empty or non-functional?**
    Oftentimes, the main Vue app script has a `defer` or `async` attribute and therefore gets added to the `deferMutations` array. Meanwhile, the `detector.js` script is the last script of `nonAsyncMutations` to be inserted back into the DOM. As explained earlier, this inserting fails. Because the queue is blocked waiting for `detector.js` to load, the main Vue app script in the `deferMutations` array is never processed and thus never executed.

**Comment by Jebble** [comment]
Identified a similar issue here today where after redirecting to an app with Cookiebot, if Vue Devtools are installed in Chrome (and not even enabled on the application), the app stops working. It fires an initial API request to our back-end but fails to handle the response and doesn't execute the `vue-router` navigation
